portia/planners/one_shot_planner.py

```
@@ -59,7 +60,40 @@ def generate_steps_or_error(
                 {"role": "user", "content": prompt},
             ],
         )
+
+        if not response.error:
+            response.error = self._validate_tools_in_response(response.steps, tool_list)
+
+        # Add LLMTool to the steps that don't have a tool_id.
+        for step in response.steps:
+            if step.tool_id is None:
+                step.tool_id = LLMTool.LLM_TOOL_ID
+
         return StepsOrError(
             steps=response.steps,
             error=response.error,
         )
+
+    def _validate_tools_in_response(self, steps: list[Step], tool_list: list[Tool]) -> str | None:
```

Is this to stop tool name hallucination? I'm not convinced we should add a point case for this given I don't see it happening regularly? If we want to address that I think we should look at developing a more complicated planner akin to the verifier agent.

---

I definitely saw it happens more than once with me. Can't reproduce it that much tbh. I agree with your sentiment, but I'd think we should be more on defensive side, rather than trusting llms. It's almost the same case as if we validate the pedantic model from the llm response.

---

IMO its the same distinction we have in the agent. The one shot agent also doesn't catch hallucinations, the verifier agent does. I just think you can add many sorts of bandaid fixes to this but we're better off fixing systemically with a smarter planner. Not going to stop merging though if you have strong conviction.

---

I really agree to your point and this why I want the planner to fail louder and clearer. This is not fixing the tool ids hallucinations.